app/controllers/application_controller.rb

```diff
@@ -5,6 +5,81 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, "fwitter secret"
   end
+  
+  get '/' do 
+  	erb :index
+  end
+  
+  get '/signup' do 
+  	if logged_in? 
+  		redirect '/tweets'
+  	else
+			erb :'users/create_user'
+		end
+	end
+	
+	post '/signup' do 
+		user = User.new(:username => params[:username], :email => params[:email], :password => params[:password])
+		if user.save && valid_user?(user)
+			session[:user_id] = user.id
+			redirect '/tweets'
+		else
+			redirect '/signup'
+		end
+	end
+	
+	get '/login' do 
+		if logged_in? 
+  		redirect '/tweets'
+  	else
+			erb :'/users/login'
+		end
+	end
+	
+	post '/login' do 
+		user = User.find_by(:username => params[:username])
+		if user && user.authenticate(params[:password])
+			session[:user_id] = user.id
+			redirect '/tweets'
+		else
+			redirect '/login'
+		end
+	end
+	
+	get '/logout' do 
+		if logged_in? 
+			session.clear
+  		redirect '/login'
+  	else
+			redirect '/'
+		end
+	end
+			
+	
+	helpers do 
+		def logged_in?
+			!!session[:user_id]
```

Comment:
It's better to rely on the successful call to `current_user` than only introspecting on the `session[:id]` value. In theory, a user might have an invalid `id` in their session which would make `logged_in?` return true even though that user would not be findable later. Also, because `current_user` is memoized, it makes sense to call it as early in the request cycle as possible, which generally is the first call `logged_in?`.

```ruby
def logged_in?
  !!current_user
end
```